File: src/cmd_parser.c
#include "cmd_parser.h"
#include "cmd.h"
#include "glib.h"
#include "utils.h"
#include <stdbool.h>
#include <stdio.h>
#include <unistd.h>

#define STR_UNQUOTED '\''
#define STR_QUOTED '"'
#define VAR_EXPAND_START '$'
#define VAR_ASSIGN '='
#define PIPE '|'
#define COMMENT '#'

static bool is_alpha(char c) {
  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
}

static bool is_numeric(char c) { return c >= 48 && c <= 57; }

static bool is_literal_char(char c) {
  return !(c == ' ' || c == '\n' || c == '$' || c == '`' || c == '<' ||
           c == '>' || c == '&' || c == STR_QUOTED || c == STR_UNQUOTED ||
           c == PIPE || c == ';');
}

static bool is_var_name_char(char c) {
  return is_alpha(c) || is_numeric(c) || c == '_';
}

static bool is_end_of_line(char c) { return c == '\n' || c == '\0'; }

static inline void parser_consume_to_end_of_line(cmd_parser *parser) {
  while (!is_end_of_line(*parser->next)) {
    parser->next++;
  }
}

#pragma clang diagnostic ignored "-Wunused-parameter"
#pragma clang diagnostic push
static void cmd_parser_err(cmd_parser *parser, char *fmt, ...) {
  va_list args;
  va_start(args, fmt);

  char full_format[BUFSIZ];
  snprintf(full_format, BUFSIZ, "parser error: %s\n", fmt);

  vfprintf(stderr, full_format, args);
  va_end(args);

  exit(1);
}
#pragma clang diagnostic pop

// cmd_parser_parse_var_expand parses a variable expansion.
//
// The cursor will be placed after the var name.
// (e.g. "$var" will be returned and the cursor will be at "/" in "$var/baz").
cmd_word_part_var *cmd_parser_parse_var_expand(cmd_parser *parser) {
  parser->next++;
  char c = *parser->next;

  cmd_word_part_var *var = malloc(sizeof(cmd_word_part_var));
  var->name = g_string_new(NULL);

  // Check for special var names.
  if (c == '!' || c == '?') {
    var->name = g_string_append_c(var->name, c);
    parser->next++;
    return var;
  }

  while (c != 0) {
    c = *parser->next;

    if (!is_var_name_char(c)) {
      break;
    }

    var->name = g_string_append_c(var->name, c);
    parser->next++;
  }

  return var;
}

// cmd_parser_parse_word_literal parses a word literal.
//
// The cursor will be placed after at the last character of the literal
// (e.g. "foo" will be returned and cursor will be at ' ' in "foo bar").
GString *cmd_parser_parse_word_literal(cmd_parser *parser) {
  GString *literal = NULL;

  while (*parser->next != '\0') {
    char c = *parser->next;

    if (parser->in_sub && c == ')') {
      return literal;
    }

    if (is_literal_char(c)) {
      if (literal == NULL) {
        literal = g_string_new(NULL);
      }

      literal = g_string_append_c(literal, c);
    } else if (c == ' ' || c == '\n') {
      return literal;
    } else if (c == ';') {
      return literal;
    } else {
      cmd_parser_err(parser, "word_literal: unexpected character %c", c);
      return NULL;
    }

    parser->next++;
  }

  return literal;
}

// cmd_parser_parse_str_literal parses a string literal.
//
// The cursor will be placed at the end of the literal
// (e.g. "foo" will be returned and the cursor will be at '$' in "foo$bar").
cmd_word_part_str_part *cmd_parser_parse_str_literal(cmd_parser *parser,
                                                     bool (*predicate)(char)) {
  cmd_word_part_str_part *part = malloc(sizeof(cmd_word_part_str_part));
  part->type = CMD_WORD_PART_STR_PART_TYPE_LITERAL;
  part->value.literal = g_string_new(NULL);

  while (*parser->next != '\0') {
    char c = *parser->next;

    if (predicate(c)) {
      part->value.literal = g_string_append_c(part->value.literal, c);
    } else {
      break;
    }

    parser->next++;
  }

  return part;
}

static bool is_str_unquoted_lit_char(char c) { return c != STR_UNQUOTED; }

// cmd_parser_parse_str_unquoted parses an unquoted string.
//
// The cursor will be placed after the string
// (e.g. "'foo'" will be returned and the cursor will be at ' ' in "'foo' bar").
cmd_word_part_str *cmd_parser_parse_str_unquoted(cmd_parser *parser) {
  parser->next++;

  cmd_word_part_str *res = malloc(sizeof(cmd_word_part_str));
  res->quoted = false;
  res->parts = g_list_append(
      NULL, cmd_parser_parse_str_literal(parser, is_str_unquoted_lit_char));

  parser->next++;

  return res;
}

static bool is_str_quoted_lit_char(char c) {
  return is_literal_char(c) || c == ' ' || c == ';';
}

// cmd_parser_parse_sub parses a command or process substitution.
//
// The cursor will be placed after the sub.
cmd *cmd_parser_parse_sub(cmd_parser *parser) {
  if (!(*parser->next == '$' || *parser->next == '<') ||
      *(parser->next + 1) != '(') {
    cmd_parser_err(parser, "unexpected char in cmd sub: %s", parser->next);
  }

  parser->next += 2;

  bool was_in_sub = parser->in_sub;
  parser->in_sub = true;
  cmd *cmd = cmd_parser_parse(parser, parser->next);
  parser->in_sub = was_in_sub;

  return cmd;
}

// cmd_parser_parse_str_quoted parses a quoted string.
//
// The cursor will be placed after the string.
// (e.g. '"foo$bar"' will be returned and the cursor will be at ' ' in
// '"foo$bar" baz').
cmd_word_part_str *cmd_parser_parse_str_quoted(cmd_parser *parser) {
  parser->next++;

  cmd_word_part_str *res = malloc(sizeof(cmd_word_part_str));
  res->quoted = true;
  res->parts = NULL;

  while (*parser->next != '\0') {
    char c = *parser->next;

    if (is_str_quoted_lit_char(c)) {
      cmd_word_part_str_part *part =
          cmd_parser_parse_str_literal(parser, is_str_quoted_lit_char);

      res->parts = g_list_append(res->parts, part);
    } else if (c == VAR_EXPAND_START) {
      cmd_word_part_var *var = cmd_parser_parse_var_expand(parser);

      cmd_word_part_str_part *part = malloc(sizeof(cmd_word_part_str_part));
      part->type = CMD_WORD_PART_STR_PART_TYPE_VAR;
      part->value.var = var;

      res->parts = g_list_append(res->parts, part);
    } else if (c == STR_QUOTED) {
      parser->next++;

      break;
    } else {
      cmd_parser_err(parser, "str_quoted: unexpected char: %c", c);
    }
  }

  return res;
}

// cmd_parser_parse_str parses a string.
//
// The cursor will be placed after the string.
// (e.g. "'foo'" will be returned and the cursor will be at ' ' in "'foo' bar").
cmd_word_part_str *cmd_parser_parse_str(cmd_parser *parser) {
  switch (*parser->next) {
  case STR_UNQUOTED: {
    return cmd_parser_parse_str_unquoted(parser);
  }

  case STR_QUOTED: {
    return cmd_parser_parse_str_quoted(parser);
  }

  default: {
    cmd_parser_err(parser, "parse_str: unexpected char in string: %c",
                   *parser->next);
  }
  }

  return NULL;
}

// cmd_parser_parse_word parses a word.
//
// The cursor will be placed after the word.
// (e.g. "foo" will be returned and the cursor will be at ' ' in "foo bar").
cmd_word *cmd_parser_parse_word(cmd_parser *parser) {
  cmd_word *word = malloc(sizeof(cmd_word));
  word->parts = NULL;

  while (*parser->next != '\0') {
    char c = *parser->next;

    if (c == COMMENT) {
      parser_consume_to_end_of_line(parser);
      return word;
    }

    if (c == ' ' || c == '\n' || c == ';' || (parser->in_sub && c == ')')) {
      return word;
    }

    // Check if this is a command sub.
    if (c == VAR_EXPAND_START && *(parser->next + 1) == '(') {
      cmd_word_part_value val = {
          .cmd_sub = cmd_parser_parse_sub(parser),
      };
      cmd_word_part *part = cmd_word_part_new(CMD_WORD_PART_TYPE_CMD_SUB, val);

      word->parts = g_list_append(word->parts, part);

      continue;
    }

    // Check if this is a proc sub.
    if (c == '<' && *(parser->next + 1) == '(') {
      cmd_word_part_value val = {
          .proc_sub = cmd_parser_parse_sub(parser),
      };
      cmd_word_part *part = cmd_word_part_new(CMD_WORD_PART_TYPE_PROC_SUB, val);

      word->parts = g_list_append(word->parts, part);

      continue;
    }

    if (is_literal_char(c)) {
      cmd_word_part_value val = {
          .literal = cmd_parser_parse_word_literal(parser),
      };
      cmd_word_part *part = cmd_word_part_new(CMD_WORD_PART_TYPE_LIT, val);

      word->parts = g_list_append(word->parts, part);
    } else if (c == STR_UNQUOTED || c == STR_QUOTED) {
      cmd_word_part_value val = {
          .str = cmd_parser_parse_str(parser),
      };
      cmd_word_part *part = cmd_word_part_new(CMD_WORD_PART_TYPE_STR, val);

      word->parts = g_list_append(word->parts, part);
    } else if (c == VAR_EXPAND_START) {
      cmd_word_part_value val = {
          .var = cmd_parser_parse_var_expand(parser),
      };
      cmd_word_part *part = cmd_word_part_new(CMD_WORD_PART_TYPE_VAR, val);

      word->parts = g_list_append(word->parts, part);
    } else {
      cmd_parser_err(parser, "parse_word: unexpected character %c", c);
    }
  }

  return word;
}

cmd_parser *cmd_parser_new(void) {
  cmd_parser *parser = malloc(sizeof(cmd_parser));
  parser->in_sub = false;

  return parser;
}

void cmd_parser_set_next(cmd_parser *parser, char *next) {
  parser->in_sub = false;
  parser->next = next;
}

// cmd_parser_parse parses the provided input and returns an executable cmd*.
cmd *cmd_parser_parse(cmd_parser *parser, char *input) {
  if (input == NULL || *input == 0) {
    return NULL;
  }

  cmd *res = cmd_new();

  parser->next = input;

  bool can_set_vars = true;
  while (*parser->next != '\0') {
    char c = *parser->next;

    while (c == ' ') {
      c = *++parser->next;
    }

    if (c == COMMENT) {
      parser_consume_to_end_of_line(parser);
      return res;
    }

    if (c == '\n' || c == ';' || (parser->in_sub && c == ')')) {
      parser->next++;
      return res;
    }

    // Check if this is a var assignment.
    if (can_set_vars && is_literal_char(c)) {
      // Peek to the ' '
      char *space_ch = strstr(parser->next, " ");

      // If that failed, try '\n'
      if (space_ch == NULL) {
        space_ch = strstr(parser->next, "\n");
      }

      // If that failed, try NULL
      if (space_ch == NULL) {
        space_ch = strstr(parser->next, "\0");
      }

      if (space_ch != NULL) {
        size_t space_index = (size_t)(space_ch - parser->next);

        // Peek to the '='.
        char *var_assign_ch = strnstr(parser->next, "=", space_index - 1);
        if (var_assign_ch != NULL) {
          size_t var_assign_index = (size_t)(var_assign_ch - parser->next);

          // Grab the name.
          char *name = malloc((var_assign_index + 1) * sizeof(char));
          memcpy(name, parser->next, var_assign_index);
          name[var_assign_index] = 0;

          // Read the value as the next word on the other side of the '='.
          parser->next = var_assign_ch + 1;
          cmd_word *value = cmd_parser_parse_word(parser);

          cmd_var_assign *var = malloc(sizeof(cmd_var_assign));
          var->name = name;
          var->value = value;

          cmd_part *part = malloc(sizeof(cmd_part));
          part->type = CMD_PART_TYPE_VAR_ASSIGN;
          part->value.var_assign = var;

          res->parts = g_list_append(res->parts, part);

          continue;
        }
      }
    }

    // Check if this is a word.
    if (is_literal_char(c) || c == STR_UNQUOTED || c == STR_QUOTED ||
        c == VAR_EXPAND_START || c == '<') {
      can_set_vars = false;

      cmd_part *part = malloc(sizeof(cmd_part));
      part->type = CMD_PART_TYPE_WORD;
      part->value.word = cmd_parser_parse_word(parser);

      res->parts = g_list_append(res->parts, part);
      continue;
    }

    // Check if this is a background proc or an AND.
    if (c == '&') {
      parser->next++;

      if (*parser->next == '&') {
        can_set_vars = true;

        parser->next++;

        // Read everything to the right of the AND as its own cmd.
        cmd *and_cmd = cmd_parser_parse(parser, parser->next);

        cmd_part *part = malloc(sizeof(cmd_part));
        part->type = CMD_PART_TYPE_AND;
        part->value.and_cmd = and_cmd;

        res->parts = g_list_append(res->parts, part);
      } else {
        giveup("parse: background procs not implemented");
      }

      continue;
    }

    // Check if this is a pipe or an OR.
    if (c == PIPE) {
      can_set_vars = true;

      parser->next++;

      if (*parser->next == '|') {
        parser->next++;

        // Read everything to the right of the OR as its own cmd.
        cmd *or_cmd = cmd_parser_parse(parser, parser->next);

        cmd_part *part = malloc(sizeof(cmd_part));
        part->type = CMD_PART_TYPE_OR;
        part->value.or_cmd = or_cmd;

        res->parts = g_list_append(res->parts, part);
      } else {
        // Read everything to the right of the pipe as its own cmd.
        cmd *piped_cmd = cmd_parser_parse(parser, parser->next);

        cmd_part *part = malloc(sizeof(cmd_part));
        part->type = CMD_PART_TYPE_PIPE;
        part->value.piped_cmd = piped_cmd;

        res->parts = g_list_append(res->parts, part);
      }

      continue;
    }

    cmd_parser_err(parser, "parse: unexpected char %c", c);
  }

  return res;
}

cmd *cmd_parser_parse_next(cmd_parser *parser) {
  return cmd_parser_parse(parser, parser->next);
}

File: src/cmd_parser.zig
const std = @import("std");
const cmd = @import("cmd.zig");

const STR_UNQUOTED = '\'';
const STR_QUOTED = '"';
const VAR_EXPAND_START = '$';
const VAR_ASSIGN = '=';
const PIPE = '|';
const COMMENT = '#';

pub const Error = error{
    OutOfInput,
    NotStarted,
};

pub const CmdParser = struct {
    const Self = @This();

    allocator: std.mem.Allocator,

    buf_offset: usize,
    buf: []u8,
    in_sub: bool,
    started: bool,

    pub fn init(allocator: std.mem.Allocator, buf: []u8) Self {
        return .{
            .allocator = allocator,
            .buf_offset = 0,
            .buf = buf,
            .in_sub = false,
            .started = false,
        };
    }

    /// parseVarExpand parses a variable expansion.
    ///
    /// The cursor will be placed after the var name.
    /// (e.g. "$var" will be returned and the cursor will be at "/" in "$var/baz").
    fn parseVarExpand(self: *Self) !*cmd.CmdWordPartVar {
        const part = try self.allocator.create(cmd.CmdWordPartVar);

        var ch = try self.next();

        // Check for special var names.
        if (ch == '!' or ch == '?') {
            part.name = try self.take(1);
            return part;
        }

        // Keep track of how many characters past the head of buf we've looked.
        //
        // We'll use this to construct the name and actually update the buf later.
        var i: u32 = 0;

        while (true) {
            // If this is a non-name character, we're done!
            if (!isVarNameChar(ch)) {
                break;
            }

            i += 1;
            ch = try self.peek(i);
        }

        part.name = try self.take(i);
        return part;
    }

    /// parseWordLiteral parses a word literal.
    ///
    /// The cursor will be placed after at the last character of the literal
    /// (e.g. "foo" will be returned and cursor will be at ' ' in "foo bar").
    fn parseWordLiteral(self: *Self) Error![]u8 {
        // Keep track of how many characters past the head of buf we've looked.
        //
        // We'll use this to construct the name and actually update the buf later.
        var i: u32 = 0;

        var ch = try self.curr();
        while (true) {
            // Check if we're done with the current substitution (if any).
            if (self.in_sub and ch == ')') {
                break;
            }

            // Check if we've reached a terminating character.
            if (ch == ' ' or ch == '\n' or ch == ';') {
                break;
            }

            // If this isn't a literal character, then report an error.
            if (!isLiteralChar(ch)) {
                self.onErr("parseWordLiteral: unexpected character {s}", .{ch});
            }

            i += 1;
            ch = self.peek(i) catch break;
        }

        return try self.take(i);
    }

    /// parseStringLiteral parses a string literal.
    ///
    /// The cursor will be placed at the end of the literal
    /// (e.g. "foo" will be returned and the cursor will be at '$' in "foo$bar").
    fn parseStringLiteral(self: *Self, predicate: *const fn (u8) bool) !*cmd.CmdWordPartStrPart {
        const part = try self.allocator.create(cmd.CmdWordPartStrPart);

        var i: u32 = 0;
        var ch = try self.curr();
        while (true) {
            if (!predicate(ch)) {
                break;
            }

            i += 1;
            ch = try self.peek(i);
        }

        part.literal = try self.take(i);
        return part;
    }

    /// parseStrUnquoted parses an unquoted string.
    ///
    /// The cursor will be placed after the string
    /// (e.g. "'foo'" will be returned and the cursor will be at ' ' in "'foo' bar").
    fn parseStrUnquoted(self: *Self) !*cmd.CmdWordPartStr {
        _ = try self.next();

        const part = try self.allocator.create(cmd.CmdWordPartStr);
        part.quoted = false;
        try part.parts.append(try self.parseStringLiteral(isStrUnquotedLitChar));

        _ = try self.next();

        return part;
    }

    /// parseSub parses a command or process substitution.
    ///
    /// The cursor will be placed after the sub.
    fn parseSub(self: *Self) anyerror!*cmd.Cmd {
        const subPrefix = try self.next();
        if (!(subPrefix == '$' or subPrefix == '<')) {
            self.onErr("parseSub: unexpected char in cmd sub: {s}", .{subPrefix});
        }

        const openingBrace = try self.next();
        if (openingBrace != '(') {
            self.onErr("parseSub: unexpected char in cmd sub: {s}", .{openingBrace});
        }

        // Create a new parser that will parse the subexpression using the current buffer position.
        var subParser = Self.init(self.allocator, self.buf[self.buf_offset..]);
        subParser.in_sub = true;
        const sub = try subParser.parse();

        // Skip ahead however many characters the subexpression parsing consumed.
        _ = try self.take(subParser.buf_offset);

        return sub;
    }

    /// cmd_parser_parse_str_quoted parses a quoted string.
    ///
    /// The cursor will be placed after the string.
    /// (e.g. '"foo$bar"' will be returned and the cursor will be at ' ' in
    /// '"foo$bar" baz').
    fn parseStrQuoted(self: *Self) !*cmd.CmdWordPartStr {
        _ = try self.next();

        const res = try self.allocator.create(cmd.CmdWordPartStr);
        res.* = cmd.CmdWordPartStr.init(self.allocator, true);

        var ch = try self.curr();
        while (true) {
            if (isStrQuotedLitChar(ch)) {
                try res.parts.append(try self.parseStringLiteral(isStrQuotedLitChar));
            } else if (ch == VAR_EXPAND_START) {
                const part = try self.allocator.create(cmd.CmdWordPartStrPart);
                part.variable = try self.parseVarExpand();

                try res.parts.append(part);
            } else if (ch == STR_QUOTED) {
                _ = try self.next();
            } else {
                self.onErr("parseStrQuoted: unexpected char: {s}", .{ch});
            }

            ch = try self.curr();
        }

        return res;
    }

    /// cmd_parser_parse_str parses a string.
    ///
    /// The cursor will be placed after the string.
    /// (e.g. "'foo'" will be returned and the cursor will be at ' ' in "'foo' bar").
    fn parseString(self: *Self) anyerror!*cmd.CmdWordPartStr {
        const ch = try self.curr();
        switch (ch) {
            STR_UNQUOTED => {
                return self.parseStrUnquoted();
            },
            STR_QUOTED => {
                return self.parseStrQuoted();
            },
            else => {
                self.onErr("parseStr: unexpected char in string: {s}", .{ch});
                unreachable;
            },
        }
    }

    /// cmd_parser_parse_word parses a word.
    ///
    /// The cursor will be placed after the word.
    /// (e.g. "foo" will be returned and the cursor will be at ' ' in "foo bar").
    fn parseWord(self: *Self) !*cmd.CmdWord {
        const word = try self.allocator.create(cmd.CmdWord);
        word.* = cmd.CmdWord.init(self.allocator);

        var ch = try self.curr();
        while (true) {
            if (ch == COMMENT) {
                try self.consumeToEndOfLine();
                return word;
            }

            if (ch == ' ' or ch == '\n' or ch == ';' or (self.in_sub and ch == ')')) {
                return word;
            }

            var part = try self.allocator.create(cmd.CmdWordPart);
            part.* = blk: {
                // Check if this is a command sub.
                if (ch == VAR_EXPAND_START and try self.peek(1) == '(') {
                    break :blk cmd.CmdWordPart{
                        .cmd_sub = try self.parseSub(),
                    };
                }
                // Check if this is a proc sub.
                else if (ch == '<' and (try self.peek(1)) == '(') {
                    break :blk cmd.CmdWordPart{
                        .proc_sub = try self.parseSub(),
                    };
                }
                // Check if this is word literal.
                else if (isLiteralChar(ch)) {
                    break :blk cmd.CmdWordPart{
                        .literal = try self.parseWordLiteral(),
                    };
                }
                // Check if this is a string.
                else if (ch == STR_UNQUOTED or ch == STR_QUOTED) {
                    break :blk cmd.CmdWordPart{
                        .str = try self.parseString(),
                    };
                }
                // Check if this is a var expansion.
                else if (ch == VAR_EXPAND_START) {
                    break :blk cmd.CmdWordPart{
                        .variable = try self.parseVarExpand(),
                    };
                } else {
                    self.onErr("parseWord: unexpected character: {s}", .{ch});
                }
            };

            try word.parts.append(part);
            ch = self.curr() catch break;
        }

        return word;
    }

    /// parse parses the init'd input and returns an executable cmd*.
    pub fn parse(self: *Self) anyerror!*cmd.Cmd {
        const res = try self.allocator.create(cmd.Cmd);
        res.* = cmd.Cmd.init(self.allocator);

        var canSetVars = true;
        var ch = try self.curr();

        while (true) {
            if (isEndOfLine(ch)) {
                break;
            }

            while (ch == ' ') {
                ch = self.next() catch break;
            }

            if (ch == COMMENT) {
                self.consumeToEndOfLine() catch break;
                break;
            }

            if (ch == '\n' or ch == ';' or (self.in_sub and ch == ')')) {
                _ = self.next() catch break;
                return res;
            }

            try res.parts.append(blk: {
                // Check if this is a literal word.
                if (isLiteralChar(ch) or ch == STR_UNQUOTED or ch == STR_QUOTED or ch == VAR_EXPAND_START or ch == '<') {
                    var word = try self.parseWord();

                    // Check if this is a var assignment.
                    if (canSetVars and word.parts.items.len == 1 and word.parts.items[0].literal.len > 0 and self.curr() catch ' ' == '=') {
                        _ = try self.next();

                        const varAssignment = try self.allocator.create(cmd.CmdVarAssign);
                        varAssignment.name = word.parts.items[0].literal;
                        varAssignment.value = try self.parseWord();

                        break :blk .{ .varAssign = varAssignment };
                    }

                    break :blk .{ .word = word };
                }

                if (ch == '&') {
                    ch = try self.next();
                    if (ch == '&') {
                        canSetVars = true;

                        _ = try self.next();

                        break :blk .{ .andCmd = try self.parse() };
                    } else {
                        self.onErr("parse: background procs not implemented", .{});
                    }
                }

                if (ch == PIPE) {
                    canSetVars = true;

                    ch = try self.next();
                    if (ch == PIPE) {
                        _ = try self.next();

                        break :blk .{ .orCmd = try self.parse() };
                    }

                    break :blk .{ .pipedCmd = try self.parse() };
                }

                self.onErr("parse: unexpected char {s}", .{ch});
                unreachable;
            });

            ch = self.curr() catch break;
        }

        return res;
    }

    fn consumeToEndOfLine(self: *Self) !void {
        var ch = try self.curr();
        while (!isEndOfLine(ch)) {
            _ = try self.next();
        }
    }

    fn next(self: *Self) Error!u8 {
        self.buf_offset += 1;
        return self.buf[self.buf_offset];
    }

    /// take consumes n many characters from the input buffer.
    ///
    /// Notes:
    ///   - The result will contain the current character.
    ///   - n will be clamped so that the returned slice never exceeds the length of the buffer.
    fn take(self: *Self, n: usize) Error![]u8 {
        var newOffset = self.buf_offset + n;
        if (newOffset >= self.buf.len) {
            newOffset = self.buf.len;
        }

        const result = self.buf[self.buf_offset..newOffset];
        self.buf_offset = newOffset;

        return result;
    }

    fn curr(self: Self) Error!u8 {
        return self.peek(0);
    }

    fn peek(self: Self, offset: usize) Error!u8 {
        var effectiveOffset = self.buf_offset + offset;

        if (effectiveOffset >= 0 and effectiveOffset < self.buf.len) {
            return self.buf[effectiveOffset];
        } else {
            return Error.OutOfInput;
        }
    }

    fn onErr(_: *Self, _: []const u8, _: anytype) noreturn {
        // std.log.err(fmt, args);
        std.os.exit(1);
    }
};

fn isAlpha(ch: u8) bool {
    return (ch >= 'a' and ch <= 'z') or (ch >= 'A' and ch <= 'Z');
}

fn isNumeric(ch: u8) bool {
    return ch >= 48 and ch <= 57;
}

fn isLiteralChar(ch: u8) bool {
    return !(ch == ' ' or ch == '\n' or ch == '$' or ch == '`' or ch == '<' or
        ch == '>' or ch == '&' or ch == STR_QUOTED or ch == STR_UNQUOTED or
        ch == PIPE or ch == ';');
}

fn isVarNameChar(ch: u8) bool {
    return isAlpha(ch) or isNumeric(ch) or ch == '_';
}

fn isEndOfLine(ch: u8) bool {
    return ch == '\n' or ch == 0;
}

fn isStrUnquotedLitChar(ch: u8) bool {
    return ch != STR_UNQUOTED;
}

fn isStrQuotedLitChar(ch: u8) bool {
    return isLiteralChar(ch) or ch == ' ' or ch == ';';
}
